File: DailyChallenges/CPP/LeetCode_1498.cpp
class Solution {
public:
    int numSubseq(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        
        int mod = 1e9+7, n = nums.size();
        vector<int> pows(n,1); // do the precomputation
        for (int i=1; i<n; ++i)
            pows[i] = pows[i-1] * 2 % mod;
        
        int l = 0, r = n-1, res = 0;
        while (l <= r) { // take l, others eith choose or not 
            if (nums[l] + nums[r] > target) 
                r--;
            else
                res = (res + pows[r - l++]) % mod;
        }
        return res;
    }
};

File: DailyChallenges/Scala/LeetCode_1498.scala
object Solution {
    def numSubseq(nums: Array[Int], target: Int): Int = {
        val snums = nums.sorted
        val pows = Array.fill[Int](nums.length-1)(0).scanLeft(1)(_ * 2 % 1000000007 + _)
        
        def helper(l: Int, r: Int, res: Int): Int = 
            if (l > r) res
            else if (snums(l) + snums(r) > target) 
                helper(l, r-1, res)
            else
                helper(l+1, r, (res + pows(r-l)) % 1000000007)

        helper(0, nums.length-1, 0)
    }
}
